### PR TITLE
MDC logging bug fixes

### PR DESCRIPTION
**Pull request by pierre**
See commit by commit for bug descriptions.

### CODE REVIEW

File: payment/src/main/java/org/killbill/billing/payment/dispatcher/CallableWithRequestData.java [hunk 1]
```diff
@@ -72,7 +72,8 @@ public T call() throws Exception {
             UUIDs.setRandom(null);
```
Comment (sbrossie) [inline payment/src/main/java/org/killbill/billing/payment/dispatcher/CallableWithRequestData.java]: Unrelated to your change but this whole `finally` block is a crazy amount of thread reset state!

File: payment/src/main/java/org/killbill/billing/payment/dispatcher/CallableWithRequestData.java [hunk 2]
```diff
@@ -72,7 +72,8 @@ public T call() throws Exception {
...
             ThreadContext.unbindSecurityManager();
             ThreadContext.unbindSubject();
-            MDC.clear();
+            // Reset the MDC as before the call, in case the plugin fiddle'd with it
+            MDC.setContextMap(mdcContextMap);
```
Comment (sbrossie) [inline payment/src/main/java/org/killbill/billing/payment/dispatcher/CallableWithRequestData.java]: This 'plugin' thread is going back to the pool - don;t we want to clear the state completely - also how do we know this was the previous state?

File: payment/src/main/java/org/killbill/billing/payment/api/DefaultPaymentApi.java [hunk 3]
```diff
@@ -898,9 +897,9 @@ public Payment createChargebackReversalWithPaymentControl(final Account account,
         PaymentTransaction paymentTransaction = null;
         PaymentApiException exception = null;
         try {
+            final InternalCallContext internalCallContext = internalCallContextFactory.createInternalCallContext(account.getId(), callContext);
```
Comment (sbrossie) [inline payment/src/main/java/org/killbill/billing/payment/api/DefaultPaymentApi.java]: Very tricky, i had to read this diff several times to realize this issue has to do with the ordering of creating the context  - and calling `populateMDCContext` -- versus doing the first log line. I am wondering if there are not more places where we have this issue, and how we gonna maintain this ?

Comment (pierre) [reply payment/src/main/java/org/killbill/billing/payment/api/DefaultPaymentApi.java]: > Very tricky, i had to read this diff several times to realize this issue has to do with the ordering of creating the context - and calling `populateMDCContext` -- versus doing the first log line.

That's why I hinted you at looking at my individual commit messages 😃 

> I am wondering if there are not more places where we have this issue, and how we gonna maintain this ?

Yup, I was wondering the same thing unfortunately. The `InternalCallContextFactory` is really **the** centralized place where we resolve the record ids for the log lines, so it kind of makes sense to populate the MDC here. Also, one of the first things we do going from API to internal code is to create the `internal` contexts. So, everything points to the fact that it's the right place/moment to do this sort of things and we are fairly consistent. But, it's definitively not bullet proof...

Sorry, I don't have a good answer here 🙁 I think we need to be due diligent when writing new log lines and inspect them manually. Alternatively, we could try to add some preconditions in the code, but it's non trivial (would have to facade all SLF4J calls probably).

File: payment/src/main/java/org/killbill/billing/payment/dispatcher/CallableWithRequestData.java (hunk 2 above)
Comment (pierre) [reply payment/src/main/java/org/killbill/billing/payment/dispatcher/CallableWithRequestData.java]: Absolutely - fixed.

**Pull request merged**